Lib/pathlib.py

```diff
@@ -1214,45 +1214,49 @@ def expanduser(self):
-    def _walk(self, top_down, on_error, follow_symlinks):
-        # We may not have read permission for self, in which case we can't
-        # get a list of the files the directory contains. os.walk
-        # always suppressed the exception then, rather than blow up for a
-        # minor reason when (say) a thousand readable directories are still
-        # left to visit. That logic is copied here.
-        try:
-            scandir_it = self._scandir()
-        except OSError as error:
-            if on_error is not None:
-                on_error(error)
-            return
-
-        with scandir_it:
-            dirnames = []
-            filenames = []
-            for entry in scandir_it:
-                try:
-                    is_dir = entry.is_dir(follow_symlinks=follow_symlinks)
-                except OSError:
-                    # Carried over from os.path.isdir().
-                    is_dir = False
-
-                if is_dir:
-                    dirnames.append(entry.name)
-                else:
-                    filenames.append(entry.name)
-
-        if top_down:
-            yield self, dirnames, filenames
-
-        for dirname in dirnames:
-            dirpath = self._make_child_relpath(dirname)
-            yield from dirpath._walk(top_down, on_error, follow_symlinks)
+        stack = [self]
+
+        while stack:
+            path = stack.pop()
+            if isinstance(path, tuple):
+                yield path
+                continue
+
+            # We may not have read permission for self, in which case we can't
+            # get a list of the files the directory contains. os.walk
+            # always suppressed the exception then, rather than blow up for a
+            # minor reason when (say) a thousand readable directories are still
+            # left to visit. That logic is copied here.
+            try:
+                scandir_it = path._scandir()
+            except OSError as error:
+                if on_error is not None:
+                    on_error(error)
+                continue
+
+            with scandir_it:
+                dirnames = []
+                filenames = []
+                for entry in scandir_it:
+                    try:
+                        is_dir = entry.is_dir(follow_symlinks=follow_symlinks)
+                    except OSError:
+                        # Carried over from os.path.isdir().
+                        is_dir = False
+
+                    if is_dir:
+                        dirnames.append(entry.name)
+                    else:
+                        filenames.append(entry.name)
+
+            if top_down:
+                yield path, dirnames, filenames
+            else:
+                stack.append((path, dirnames, filenames))
 
-        if not top_down:
-            yield self, dirnames, filenames
+            for dirname in reversed(dirnames):
+                dirpath = path._make_child_relpath(dirname)
+                stack.append(dirpath)
```

Comment:
```suggestion
            stack.extend(path._make_child_relpath(dirname) for dirname in reversed(dirnames))
```

Or perhaps

```suggestion
            stack += [path._make_child_relpath(dirname) for dirname in reversed(dirnames)]
```

Might be faster, and seems no less clear.

Reply:
In my experience, listcomp is faster in such cases but the primary benefit of this is how clear it is.